CC-2669 Allow id and version to be passed to register; add modes

rayokota opened the pull request:
The two main changes are the ability to pass an id and version during schema registration, and the support for modes for a subject (or a set of subjects if the subject name used for the mode ends with an asterisk).  The three modes are read-write (the default), read-only, and import.  Passing an id or version during schema registration is only allowed during import mode.

avocader commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/MockSchemaRegistryClient.java:
```diff
@@ -43,6 +43,7 @@
   private final Map<String, Map<Integer, Schema>> idCache;
   private final Map<String, Map<Schema, Integer>> versionCache;
   private final Map<String, String> compatibilityCache;
+  private String mode = "READWRITE";
```
nit: make it final?

avocader replied:
Never mind, I see that you modify it later.

avocader commented on core/src/main/java/io/confluent/kafka/schemaregistry/exceptions/OperationNotPermittedException.java:
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015 Confluent Inc.
```
nit: year

rayokota replied:
Ack.

rayokota commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java:
```diff
@@ -141,6 +143,11 @@ private int registerAndGetId(String subject, Schema schema)
     return restService.registerSchema(schema.toString(), subject);
   }
 
+  private int registerAndGetId(String subject, Schema schema, int id)
```
I'm going to leave this for now as the underlying API also returns an ID.

rayokota commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java:
```diff
@@ -163,19 +170,31 @@ private int getIdFromRegistry(String subject, Schema schema)
     final Integer cachedId = schemaIdMap.get(schema);
     if (cachedId != null) {
+      if (id >= 0 && id != cachedId) {
+        throw new IllegalStateException("Schema already registered with id "
```
There are other IllegalStateExceptions being thrown elsewhere, so Im going to leave this for now.

rayokota commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/SchemaRegistryClient.java:
```diff
@@ -57,6 +60,11 @@ public String updateCompatibility(String subject, String compatibility)
 
   public String getCompatibility(String subject) throws IOException, RestClientException;
 
+  public String setMode(String mode)
```
Good catch!  I added them.

rayokota commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/rest/entities/requests/RegisterSchemaRequest.java:
```diff
@@ -23,12 +23,23 @@
 
 public class RegisterSchemaRequest {
 
+  private int id = -1;
```
Good catch!

rayokota commented on client/src/test/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClientTest.java:
```diff
@@ -73,6 +87,30 @@ public void testRegisterSchemaCache() throws Exception {
 
+  @Test
+  public void testRegisterSchemaCacheWithId() throws Exception {
+    CachedSchemaRegistryClient client = new CachedSchemaRegistryClient(restService, 20, new
```
Good catch!  Those constants appeared after I merged from master, I used them as you suggested.

rayokota commented on client/src/test/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClientTest.java:
```diff
@@ -223,6 +261,44 @@ public void testDeleteVersionCache() throws Exception {
 
   @Test
+  public void testSetMode() throws Exception {
+    CachedSchemaRegistryClient client = new CachedSchemaRegistryClient(restService, 20,  new
```
Ack.

rayokota commented on client/src/test/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClientTest.java:
```diff
@@ -223,6 +261,44 @@ public void testDeleteVersionCache() throws Exception {
+
+  @Test
+  public void testGetMode() throws Exception {
+    CachedSchemaRegistryClient client = new CachedSchemaRegistryClient(restService, 20,  new
```
Ack.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/exceptions/IdDoesNotMatchException.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class IdDoesNotMatchException extends SchemaRegistryException {
+
+  public IdDoesNotMatchException(String message, Throwable cause) {
```
Good suggestion.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/exceptions/OperationNotPermittedException.java:
```diff
@@ -0,0 +1,39 @@
+package io.confluent.kafka.schemaregistry.exceptions;
+
+/**
+ * Schema registry is in a mode that does not permit the requested operation.
```
I changed the description.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -238,6 +243,8 @@
       "If true, this node can participate in master election. In a multi-colo setup, turn this off "
       + "for clusters in the slave data center.";
+  protected static final String MODE_MUTABILITY_DOC =
+      "If true, the mode of this node can be changed.";
```
Good catch, I changed the description.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/resources/SubjectVersionsResource.java:
```diff
@@ -155,12 +156,14 @@ public void register(final @Suspended AsyncResponse asyncResponse,
 
-    Schema schema = new Schema(subjectName, 0, 0, request.getSchema());
+    Schema schema = new Schema(subjectName, 0, request.getId(), request.getSchema());
     int id = 0;
```
Ack.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/ModeKey.java:
```diff
@@ -0,0 +1,90 @@
+  @Override
+  public int compareTo(SchemaRegistryKey that) {
+    int compare = super.compareTo(that);
+    if (compare == 0) {
```
Ack.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/ModeKey.java:
```diff
@@ -0,0 +1,90 @@
+    if (compare == 0) {
+      ModeKey otherKey = (ModeKey) that;
+
+      // Sort by length of subject (or prefix) in descending order
```
I think it's ok as-is.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java:
```diff
@@ -830,6 +939,86 @@ private void deleteSubjectCompatibility(String subject) throws StoreException {
+      ModeValue modeValue = entry.getValue();
+      // Check for subject match
+      String prefix = modeKey.getSubject();
+      if (prefix.endsWith(ModeKey.SUBJECT_WILDCARD)) {
```
I like this suggestion.  I've added `ModeKey.isPrefix` and `ModeKey.getPrefix`.  It seems to have simplified things a bit.

rayokota commented on core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java:
```diff
@@ -0,0 +1,148 @@
+        restApp.restClient.setMode(mode).getMode());
+
+    // register a valid avro
+    String schemaString1 = AvroUtils.parseSchema(
```
Ack.

rayokota commented on core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java:
```diff
@@ -0,0 +1,148 @@
+        mode,
+        restApp.restClient.setMode(mode).getMode());
+
+    // register a valid avro
```
Ack.

rayokota commented:
Thanks for the review @C0urante and @avocader !  I'm allowing reads during IMPORT mode for now as I think there is a use case for it, namely when slowly migrating consumers over to the destination SR while new schemas are still be created in the target SR.

C0urante commented on client/src/main/java/io/confluent/kafka/schemaregistry/client/rest/entities/requests/RegisterSchemaRequest.java:
```diff
@@ -58,31 +59,18 @@ public boolean equals(Object o) {
-    }
-
-    return true;
+    return id == that.id && Objects.equals(schema, that.schema);
```
🙏 thanks for cleaning up the logic here, this method was pretty hard to read as-was!

C0urante commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -244,7 +244,7 @@
       + "for clusters in the slave data center.";
   protected static final String MODE_MUTABILITY_DOC =
-      "If true, the mode of this node can be changed.";
+      "If true, the mode of this schema registry cluster can be changed.";
```
Is this strictly true? What happens if I have two nodes with different settings for `mode.mutability`?

rayokota replied:
Good point.  I've corrected it to be `If true, this node will allow mode changes if is the master.`

C0urante replied:
Great, I really like that language.

mageshn commented on core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java:
```diff
@@ -0,0 +1,131 @@
+    String mode = "IMPORT";
+
+    // set mode to import
+    assertEquals(
```
should we test register with id? Perhaps, it make sense to mandate id to be passed in the import mode?

rayokota replied:
Yes, ID is mandated for import mode.  I added a test earlier.

mageshn commented on core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java:
```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2014 Confluent Inc.
```
nit: copyright year and license

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/ModeValue.java:
```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018 Confluent Inc.
```
nit : you might want to check the copyright in all new new files

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java:
```diff
@@ -673,6 +739,16 @@ public SchemaString get(int id) throws SchemaRegistryException {
     }
   }
 
+  public Set<String> listSubjects(String subject)
```
unit test for this?

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java:
```diff
@@ -817,19 +896,94 @@ public boolean isCompatible(String subject,
+    if (mode == Mode.IMPORT && getMode(subject) != Mode.IMPORT) {
+      // Changing to import mode requires that no schemas exist with matching subjects.
+      Set<String> matchingSubjects = listSubjects(subject);
+      if (!matchingSubjects.isEmpty()) {
```
Could there be scenarios where the subject exists with no schema versions?

rayokota replied:
For import I only care that no schema versions exist for the matching subjects.

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -110,6 +110,7 @@ public void init() throws StoreInitializationException {
       throw new StoreInitializationException(
           "Illegal state while initializing store. Store was already initialized");
     }
+    localStore.init();
```
Why is the init required here?

rayokota replied:
Currently `InMemoryCache.init` is a noop, but at one point I was trying something out where I had some code in `InMemoryCache.init` and was surprised that it was not called.  This seems like the natural place to call it.

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreMessageHandler.java:
```diff
@@ -60,8 +60,9 @@ public boolean validateUpdate(SchemaRegistryKey key, SchemaRegistryValue value)
-          // Check the old schema even if it may be deleted, as the schema may still be in use.
-          if (oldSchema != null && !oldSchema.getSchema().equals(schemaObj.getSchema())) {
+          if (oldSchema != null
+              && !oldSchema.isDeleted()
```
You're right, this check is wrong.  I don't want to check for duplicate IDs in IMPORT mode, so instead I added the following lines at the beginning of this method:
```
     if (schemaObj != null 
          && lookupCache.mode(schemaObj.getSubject(), true, Mode.READWRITE) != Mode.IMPORT) {
```

rayokota commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreMessageHandler.java:
```diff
@@ -125,7 +126,9 @@ private void handleSchemaUpdate(SchemaKey schemaKey, SchemaValue schemaObj) {
-        schemaKeys.stream().filter(v -> v.getSubject().equals(schemaObj.getSubject()))
+        schemaKeys.stream().filter(v ->
+            v.getSubject().equals(schemaObj.getSubject())
+                && v.getVersion() != schemaObj.getVersion())
```
If subjects were accidentally populated before IMPORT mode, I allow the user to delete all subjects and then set to IMPORT mode.  In that case, version 1 may have been deleted, and then version 1 may be reimported, so this code just makes sure not to tombstone the deleted version 1 as that will tombstone the new version 1 as well.

rayokota commented:
Thanks @mageshn !  I refactored the `LookupCache` as a result of your feedback.  Let me know what you think.

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/InMemoryCache.java:
```diff
@@ -174,26 +176,53 @@ public Mode mode(String subject,
+            SchemaValue value = (SchemaValue) v;
+            if (value != null && !value.isDeleted()) {
+              if (match.test(key.getSubject())) {
+                return Stream.of(key.getSubject());
```
Based on this implementation, should this method return a boolean,  perhaps called `hasSubjects` or something along those lines?

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/InMemoryCache.java:
```diff
@@ -174,26 +176,53 @@ public Mode mode(String subject,
-    while (allKeys.hasNext()) {
-      K k = allKeys.next();
+  @Override
+  public Set<String> subjects(Predicate<String> match) throws StoreException {
```
perhaps it's cleaner to keep the predicate out of the interface and make it an implementation detail?

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/InMemoryCache.java:
```diff
@@ -174,26 +176,53 @@ public Mode mode(String subject,
+  }
+
+  @Override
+  public void clearSubjects(Predicate<String> match) throws StoreException {
```
should we make this method an implementation detail than having as part of the interface?

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java:
```diff
@@ -817,19 +898,72 @@ public boolean isCompatible(String subject,
+    if (!allowModeChanges) {
+      throw new OperationNotPermittedException("Mode changes are not allowed");
+    }
+    ClearSubjectKey clearSubjectKey = new ClearSubjectKey(subject);
```
The instance can be declared inside the if condition.

mageshn commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/InMemoryCache.java:
```diff
@@ -127,4 +129,92 @@ public void schemaRegistered(SchemaKey schemaKey, SchemaValue schemaValue) {
+  }
+
+  public void clearSubjects(Predicate<String> match) throws StoreException {
+    store.entrySet().removeIf(e -> {
```
This implementation assumes that the event is triggered only by SR and removes everything from the cache. What happens when someone, intentionally/un-intentionally publishes this event when the Subject is not IMPORT ready?
If the general idea is to clean up the deletes from the cache just so that import works without too much fuzziness in the logic, would it be safer to just act on the deleted schema versions and also rename the event to call it out specifically?

rayokota replied:
Great point!  I'll modify the code to only clear out deleted schemas.

mageshn commented on core/src/test/java/io/confluent/kafka/schemaregistry/rest/RestApiModeTest.java:
```diff
@@ -0,0 +1,131 @@
+
+    // register a valid avro schema
+    int expectedIdSchema1 = 1;
+    assertEquals("Registering without id should succeed",
```
nit: I got carried away with the message. I think your testing with id but the message is incorrect :)

rayokota commented:
Thanks @mageshn !  Yes, I forgot to mention that I had added `RestApiModeTest.testImportModeWithoutId` earlier to address your point about requiring ID during import.  Also, I cleaned up the ClearSubject handling by ensuring that the schemas are deleted.  Great suggestion!
